### PR TITLE
[AutoMM] FT_Transformer refactor

### PR DESCRIPTION
**Pull request by taoyang1122**
*Issue #, if available:*

*Description of changes:*
- Codes refactor of FT_Transformer. 
- Build FT_Transformer as a single model, instead of the combination of numerical_transformer, categorical_transformer and fusion_transformer.
- Add pooled_feature mode ('cls' or 'mean') for FT_Transformer.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
@taoyang1122 have we benchmarked the performance compared to the current mainline?

File: examples/automm/tabular_dl/example_tabular.py [hunk 1]
```diff
@@ -35,8 +35,10 @@
 
 automm_hyperparameters = {
     "data.categorical.convert_to_text": False,
-    "model.names": ["categorical_transformer", "numerical_transformer", "fusion_transformer"],
-    "model.numerical_transformer.embedding_arch": ["linear"],
+    # "model.names": ["categorical_transformer", "numerical_transformer", "fusion_transformer"],
+    # "model.numerical_transformer.embedding_arch": ["linear"],
+    "model.names": ["ft_transformer_refactor"],
```
Comment (zhiqiangdon) [inline examples/automm/tabular_dl/example_tabular.py:40]: `ft_transformer_refactor ` - > `ft_transformer`. Change other places accordingly.

Comment (zhiqiangdon) [reply examples/automm/tabular_dl/example_tabular.py:40]: Need to remove commented code.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 2]
```diff
@@ -251,3 +251,27 @@ model:
     data_types:
     additive_attention: False # Whether to use lightweight additive attention, can be True, False or "auto"
     share_qv_weights: False # Whether to share weight for query and value, can be True, False or "auto"
+  ft_transformer_refactor:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:254]: Need to remove the configs of `categorical_transformer` and `numerical_transformer`?

File: multimodal/src/autogluon/multimodal/utils/model.py [hunk 3]
```diff
@@ -114,6 +116,7 @@ def select_model(
     fusion_model_name = []
     for model_name in names:
         model_config = getattr(config.model, model_name)
+        strict = getattr(model_config, "select_model_strict", True)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/model.py:119]: `getattr(model_config, "select_model_strict", True)` -> `model_config.requires_all_dtypes`. Looks unnecessary to use getattr and default value.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 4]
```diff
@@ -251,3 +251,27 @@ model:
...
+    attention_n_heads: 8
+    attention_dropout: 0.2
+    residual_dropout: 0.0
+    ffn_dropout: 0.1
+    ffn_d_hidden: 192
+    ffn_activation: "geglu"
+    head_activation: "relu"
+    normalization: "layer_norm"
+    merge: "concat"
+    select_model_strict: False
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:274]: Maybe `requires_all_dtypes ` is more intuitive?

**Review by zhiqiangdon — COMMENTED** [review]
If the refactor can reproduce the previous results, we can safely remove the old implementations such as categorical_transformer.py and numerical_transformer.py. Also need to rename the previous ft_transformer.py to other names like custom_transformer.py.

File: multimodal/src/autogluon/multimodal/utils/model.py (hunk 3 above)
Comment (taoyang1122) [reply multimodal/src/autogluon/multimodal/utils/model.py:119]: But then we need to add this attribute to all previous model configs. Just thought the default should be true. But for some structures like ft_transformer, it could be false.

File: examples/automm/kaggle_california_house_price/example_kaggle_house.py [hunk 5]
```diff
@@ -33,10 +33,8 @@ def get_parser():
 
 def get_automm_hyperparameters(mode, text_backbone, cat_as_text):
     if mode == "ft-transformer":
-        hparams = {"model.names": ["categorical_transformer",
-                                   "numerical_transformer",
-                                   "hf_text",
-                                   "fusion_transformer"],
+        hparams = {"model.names": ["ft_transformer",
+                                   "hf_text"],
```
Comment (zhiqiangdon) [inline examples/automm/kaggle_california_house_price/example_kaggle_house.py:37]: Shouldn't we keep `fusion_transformer` to fuse `ft_transformer ` and `hf_text`?

File: examples/automm/tabular_dl/run_all.sh [hunk 6]
```diff
@@ -4,8 +4,8 @@ set -e
 for dataset_name in ad al ca co ep he hi ja mi ya ye
 do
   python3 example_tabular.py --dataset_name ${dataset_name} \
-  --dataset_dir ./dataset/ \
-  --exp_dir ./result/${dataset_name} \
+  --dataset_dir ./datasets/ \
+  --exp_dir ./exp/${dataset_name} \
```
Comment (zhiqiangdon) [inline examples/automm/tabular_dl/run_all.sh:8]: Is this change necessary?

File: multimodal/src/autogluon/multimodal/utils/model.py [hunk 7]
```diff
@@ -135,13 +134,6 @@ def select_model(
     if len(selected_model_names) > 1:
         assert len(fusion_model_name) == 1
         selected_model_names.extend(fusion_model_name)
-    elif len(fusion_model_name) == 1 and hasattr(config.model, fusion_model_name[0]):
-        # TODO: Support using categorical_transformer or numerical_transformer alone without a fusion model.
-        if selected_model_names[0].lower().startswith((CATEGORICAL_TRANSFORMER, NUMERICAL_TRANSFORMER)):
-            selected_model_names.extend(fusion_model_name)
-        else:
-            # remove the fusion model's config make `config.model.names` and the keys of `config.model` consistent.
-            delattr(config.model, fusion_model_name[0])
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/model.py:144]: I think we need to keep this `else`. If only one model is selected, we need to clean up the fusion model config.

File: multimodal/tests/unittests/predictor/test_predictor.py [hunk 8]
```diff
@@ -456,41 +456,29 @@ def test_model_configs():
...
-                "d_token": 192,
-                "num_trans_blocks": 0,
-                "num_attn_heads": 4,
                 "residual_dropout": 0.0,
-                "attention_dropout": 0.2,
                 "ffn_dropout": 0.1,
-                "normalization": "layer_norm",
-                "ffn_activation": "reglu",
+                "ffn_d_hidden": 192,
+                "ffn_activation": "geglu",
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/predictor/test_predictor.py:474]: why `ffn_activation ` is changed from `reglu` to `geglu`?

File: multimodal/tests/unittests/predictor/test_predictor.py [hunk 9]
```diff
@@ -456,41 +456,29 @@ def test_model_configs():
...
-                "residual_dropout": 0.0,
+                "adapter_output_feature": 192,
+                "hidden_size": 192,
+                "n_blocks": 3,
+                "attention_n_heads": 8,
                 "attention_dropout": 0.2,
-                "ffn_dropout": 0.1,
-                "normalization": "layer_norm",
-                "ffn_activation": "reglu",
-                "head_activation": "relu",
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/predictor/test_predictor.py:473]: Looks like `head_activation ` is removed from config?

File: multimodal/tests/unittests/predictor/test_predictor.py [hunk 10]
```diff
@@ -456,41 +456,29 @@ def test_model_configs():
...
-                "data_types": ["categorical"],
-                "additive_attention": "auto",
-                "share_qv_weights": "auto",
-            },
-            "numerical_transformer": {
-                "out_features": 192,
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/predictor/test_predictor.py:479]: Looks like `out_features ` is removed from config?

File: multimodal/tests/unittests/predictor/test_predictor.py [hunk 11]
```diff
@@ -456,41 +456,29 @@ def test_model_configs():
...
                 "fusion_transformer",
             ],
-            "categorical_transformer": {
-                "out_features": 192,
+            "ft_transformer": {
+                "data_types": ["numerical", "categorical"],
+                "embedding_arch": ["linear"],
                 "d_token": 192,
-                "num_trans_blocks": 0,
-                "num_attn_heads": 4,
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/predictor/test_predictor.py:467]: Looks like `num_attn_heads ` is removed from config?

File: multimodal/tests/unittests/predictor/test_predictor.py [hunk 12]
```diff
@@ -456,41 +456,29 @@ def test_model_configs():
...
-                "residual_dropout": 0.0,
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/predictor/test_predictor.py:468]: Ditto.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 13]
```diff
@@ -251,3 +215,26 @@ model:
     data_types:
     additive_attention: False # Whether to use lightweight additive attention, can be True, False or "auto"
     share_qv_weights: False # Whether to share weight for query and value, can be True, False or "auto"
+  ft_transformer:
+    data_types:
+      - "categorical"
+      - "numerical"
+    embedding_arch:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:222]: Looks like the default value of `embedding_arch ` changes from `linear` and `relu` to `linear`.

**Review by zhiqiangdon — COMMENTED** [review]
I find some default values of ft transformer are changed. You may consider the effect of these changes.

File: multimodal/tests/unittests/predictor/test_predictor.py (hunk 11 above)
Comment (taoyang1122) [reply multimodal/tests/unittests/predictor/test_predictor.py:467]: `num_attn_heads` is for categorical_transformer.

File: multimodal/tests/unittests/predictor/test_predictor.py (hunk 12 above)
Comment (taoyang1122) [reply multimodal/tests/unittests/predictor/test_predictor.py:468]: Still has `residual_dropout ` in line 471.

File: multimodal/tests/unittests/predictor/test_predictor.py (hunk 9 above)
Comment (taoyang1122) [reply multimodal/tests/unittests/predictor/test_predictor.py:473]: In line 475

File: multimodal/tests/unittests/predictor/test_predictor.py (hunk 10 above)
Comment (taoyang1122) [reply multimodal/tests/unittests/predictor/test_predictor.py:479]: `out_features` is for categorical transformer.

File: multimodal/tests/unittests/predictor/test_predictor.py (hunk 8 above)
Comment (taoyang1122) [reply multimodal/tests/unittests/predictor/test_predictor.py:474]: `reglu` is used in categorical transformer. `geglu` is used in ft_transformer backbone.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml (hunk 13 above)
Comment (taoyang1122) [reply multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:222]: Default should be `linear`, which is the same as ft_transformer. The other `embedding_arch` are different variants.

File: examples/automm/kaggle_california_house_price/example_kaggle_house.py (hunk 5 above)
Comment (taoyang1122) [reply examples/automm/kaggle_california_house_price/example_kaggle_house.py:37]: Yes, will update

File: examples/automm/tabular_dl/run_all.sh (hunk 6 above)
Comment (taoyang1122) [reply examples/automm/tabular_dl/run_all.sh:8]: Will update

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Nice refactor!
